kafka-instrumentation/src/main/java/dev/responsive/kafka/clients/ResponsiveKafkaClientSupplier.java

```diff
@@ -0,0 +1,276 @@
+    final JmxReporter jmxReporter = new JmxReporter();
+    jmxReporter.configure(configs);
+    final MetricsContext metricsContext
+        = new KafkaMetricsContext("dev.responsive", new HashMap<>());
+    metrics = factories.createMetrics(
+        new MetricConfig(),
+        Collections.singletonList(jmxReporter),
+        Time.SYSTEM,
+        metricsContext
+    );
+    endOffsetsPoller = factories.createEndOffsetPoller(configs, metrics);
+    configured = true;
+  }
+
+  @Override
+  public Admin getAdmin(final Map<String, Object> config) {
+    return wrapped.getAdmin(config);
+  }
+
+  @Override
+  public Producer<byte[], byte[]> getProducer(final Map<String, Object> config) {
+    assert configured;
+    if (!eosV2) {
+      return wrapped.getProducer(config);
+    }
+    LOGGER.info("Creating responsive producer");
+    final String tid = threadIdFromProducerConfig(config);
+    synchronized (this) {
+      return factories.createResponsiveProducer(
+          wrapped.getProducer(config),
+          Collections.unmodifiableList(
+              Arrays.asList(
+                  addOrRefCommitListener(tid),
+                  new CloseListener(tid)
+              )
+          )
+      );
+    }
+  }
+
+  @Override
+  public Consumer<byte[], byte[]> getConsumer(final Map<String, Object> config) {
+    assert configured;
+    if (!eosV2) {
+      return wrapped.getConsumer(config);
+    }
+    LOGGER.info("Creating responsive consumer");
+    final String tid = threadIdFromConsumerConfig(config);
+    synchronized (this) {
+      // TODO: the end offsets poller call is kind of heavy for a synchronized block
+      return factories.createResponsiveConsumer(
+          wrapped.getConsumer(config),
+          List.of(
+              addOrRefCommitListener(tid),
+              endOffsetsPoller.addForThread(tid),
+              new CloseListener(tid)
+          )
+      );
+    }
+  }
+
+  private MetricPublishingCommitListener addOrRefCommitListener(final String threadId) {
+    return metricListeners.compute(threadId, (t, v) -> {
+      if (v == null) {
+        LOGGER.info("Create a new metric publishing commit listener");
+        return new ReferenceCounted<>(
+            factories.createMetricsPublishingCommitListener(metrics, threadId));
+      }
+      v.ref();
+      return v;
+    }).getVal();
+  }
+
+  @Override
+  public Consumer<byte[], byte[]> getRestoreConsumer(final Map<String, Object> config) {
+    return wrapped.getRestoreConsumer(config);
+  }
+
+  @Override
+  public Consumer<byte[], byte[]> getGlobalConsumer(final Map<String, Object> config) {
+    return wrapped.getGlobalConsumer(config);
+  }
+
+  private String threadIdFromProducerConfig(final Map<String, Object> config) {
+    final String clientId = (String) config.get(CommonClientConfigs.CLIENT_ID_CONFIG);
+    LOGGER.info("Found producer client ID {}", clientId);
+    final var regex = Pattern.compile(".*-(StreamThread-\\d+)-producer");
```

Comment:
This doesn't hold up for eos-v1 (just a note for the future in case we run into any sad clients still using it)

Reply:
Ack - will leave my eosv2 only check in for now then.